lib/improver/utilities/mathematical_operations.py

```diff
@@ -0,0 +1,228 @@
+        elif not increasing_order and direction == "positive":
+            cube = sort_coord_in_cube(cube, coord_name)
+        elif not increasing_order and direction == "negative":
+            pass
+        return cube
+
+    def process(self, cube):
+        """Integrate a specified coordinate. This is calculated by defining the
+        upper and lower bounds for the steps along a chosen coordinate
+        within the cube.
+
+        Integration is performed by firstly defining the stride as the
+        difference between the upper and lower bound. The contribution from
+        the uppermost half of the stride is calculated by multiplying the
+        upper bound value by 0.5 * stride, and the contribution
+        from the lowermost half of the stride is calculated by multiplying the
+        lower bound value by 0.5 * stride. The contribution from the
+        uppermost half of the stride and the bottom half of the stride is
+        summed.
+
+        As the coordinate is progressively integrated, the contribution of
+        each stride is cumulatively summed.
+
+        Args:
+            cube (Iris.cube.Cube):
+                The cube containing the coordinate to check.
+
+        Returns:
+            integrated_cube (Iris.cube.Cube):
+                The cube containing the result of the integration.
+                This will contain the same metadata as the input cube.
+
+        """
+        # Make coordinate monotonic in the direction desired for integration.
+        cube = self.ensure_monotonic_increase_in_chosen_direction(cube)
+
+        # Define upper and lower level cubes for the integration.
+        if self.direction_of_integration == "positive":
+            upper_bounds = cube.coord(self.coord_name_to_integrate).points[1:]
+            lower_bounds = cube.coord(self.coord_name_to_integrate).points[:-1]
+        elif self.direction_of_integration == "negative":
+            upper_bounds = cube.coord(self.coord_name_to_integrate).points[:-1]
+            lower_bounds = cube.coord(self.coord_name_to_integrate).points[1:]
+
+        upper_bounds_cube = (
+             cube.extract(
+                 iris.Constraint(
+                     coord_values={self.coord_name_to_integrate:
+                                   upper_bounds})))
+        lower_bounds_cube = (
+            cube.extract(
+                iris.Constraint(
+                    coord_values={self.coord_name_to_integrate:
+                                  lower_bounds})))
+
+        # Determine which cube to copy in order to have the most appropriate
+        # points within the coordinate that is being integrated.
+        if self.direction_of_integration == "positive":
+            integrated_cube = upper_bounds_cube.copy()
+        elif self.direction_of_integration == "negative":
+            integrated_cube = lower_bounds_cube.copy()
+        integrated_cube.data = np.zeros(lower_bounds_cube.shape)
```

Comment:
Lines 146-171 could be refactored into a function make_working_cubes that defines:
self.upper_bounds_cube
self.lower_bounds_cube
self.integrated_cube

Reply:
I'll deal with this in a separate commit.

Reply:
Now refactored as suggested.